enhancements/olm/safe-cluster-upgrades.md

```diff
@@ -0,0 +1,267 @@
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# Safe OpenShift Upgrades with Operators
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Definitions
+
+**OLM Managed Operators**: Operators that are installed, managed, and upgraded by the [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) project.
+
+**Minor OpenShift Upgrade**: When the OpenShift Cluster is upgraded to the next minor version, for example the upgrade from version 4.6 to 4.7. See [Semantic Versioning](https://semver.org/) for more details.
+
+## Summary
+
+OpenShift Cluster Admins introduce new services to their clusters in the form of OLM Managed Operators. When performing a cluster upgrade, there is no guarantee that these operators will continue to run on the upgraded cluster.
+
+OLM must make an effort to prevent cluster upgrades when it is possible to determine that one or more installed operators will not run on the next version of OpenShift.
+
+## Motivation
+
+The primary purpose of this enhancement is to define how OpenShift can protect existing services introduced by OLM Managed Operators when performing a Minor OpenShift upgrade.
+
+### Goals
+
+- Prevent Minor OpenShift Upgrades when OLM detects that an installed operator will not run on the next Minor OpenShift version.
+
+### Non-Goals
+
+- Ensure that services will not be disrupted during a cluster upgrade.
+- Guarantee that all installed operator will continue to run after a cluster upgrade.
+- Prevent Patch OpenShift Upgrades.
+
+## Proposal
+
+The [ClusterVersionOperator](https://github.com/openshift/cluster-version-operator) provides OLM with the means to prevent Minor Semantic Version cluster upgrades by setting the  `operator-lifecycle-manager` [ClusterOperator's](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md) [upgradeable condition](https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#what-should-an-operator-report-with-clusteroperator-custom-resource) to `False`.
+
+OLM currently reports that it is upgradeable as soon as the OLM Deployment is successfully rolled out. This existing logic should be updated to monitor the collection of OLM Managed Operators on cluster and update the `upgradeable condition` to reflect if any of these operators will not run on the next Minor OpenShift Version.
+
+The bulk of this enhancement focuses on defining the steps that OLM will take to determine upgrade safety based on the collection of installed operators.
+
+### Allowing Operator Authors to define supported OpenShift Versions
```

Comment:
operators that publish to the redhat operators, isv and community operator catalogs that are on cluster use a label on their bundle image to indicate which OCP versions they support (see: https://docs.engineering.redhat.com/display/CFC/Delivery). It seems redundant to have them also add a field in the CSV.

Reply:
@gallettilance thanks for sharing that delivery document. We can absolutely pursue that approach since this is an OpenShift only feature. However, there are two concerns that I would like to address before moving forward:

1. The "Max OpenShift Version" could only be inferred when the bundle image is shipped to a specific version of OpenShift (`V4.6`) or a range of OpenShift Version (`v4.5-v4.7`, where we would assume the end of the range as the Max OpenShift Version, 4.7 in this case) as all other forms are shipped to future OpenShift versions. This means that the "Max OpenShift Version" could not be inferred for a set of labels that are currently in use so OLM would still need to provide the ability for this subset of operators to specify a "Max OpenShift Version".
2. Currently, these labels simply identify which OpenShift versions they should be available on. These labels will instead prevent the Cluster from upgrading to the next version. I suspect there might be a number of operators that will unexpectedly prevent cluster upgrades if we make this change. Consider a 4.,6 cluster where a bundle is shipped labeled with the `=4.6` label and the next version of the bundle is shipped with `=4.7`. The 4.6 cluster would not be able to upgrade to 4.7 as the existing bundle would set the max OpenShift Version to 4.6. We could instead assume that the Max OpenShift Version is equal to 1 more than the provided version (Example: a bundle labeled with `=4.6` equals a "Max OpenShift Version" of `4.7`) but that would not handle the usecase where a bundle made available on 4.6 is known not to run on 4.7.

Overall, I am concerned that if we commandeer these labels make unfair assumptions about intended operator behavior.

Reply:
1. Operator authors can't modify their CSV after they are released. In both cases (label and CSV annotation) they have to use their best guess (which is why there is a deprecation mechanism in the pipeline) and that guess will be wrong exactly in the same way for both right? let me know if I misunderstood what you were trying to say
2. I think we need to distinguish "installable" and "upgradeable". In the case where I set a max version in the labels, the intention is for the version to be no longer installable on OCP versions after that max version. However, it may still be able to run and upgrade on OCP versions after that max version - if it cannot maybe this is where deprecation comes into play (given that the resolver has access to that information via properties on the bundle, it could surface that somehow). To be honest we haven't explicitely made that distinction in the pipeline but it definitely feels like a reasonable ask given this enhancement. What do you think?

Reply:
Thanks for the detailed response @gallettilance 

>1. Operator authors can't modify their CSV after they are released. In both cases (label and CSV annotation) they have to use their best guess (which is why there is a deprecation mechanism in the pipeline) and that guess will be wrong exactly in the same way for both right? let me know if I misunderstood what you were trying to say

There is no need to modify the CSV after it is released, consider the following example:
- Operator Foo has been developed for and tested against a 4.7 cluster.
- The Author sets the maxOpenShiftVersion annotation on Operator Foo's CSV to 4.7
- OpenShift 4.8 becomes available.
- Clusters with Operator Foo are unable to upgrade to 4.8
- A new version of Operator Foo is released in the existing 4.7 index / channel. This version of Operator Foo has been tested on OpenShift Version 4.8 and has had the maxOpenShiftVersion annotation set to 4.8.
- (Optional Step) All earlier versions of Operator Foo are marked as deprecated. (unsure if this is needed).
- Cluster Admins can now upgrade Operator Foo to the latest version, unblocking Cluster Upgrades.

>  I think we need to distinguish "installable" and "upgradeable"

Agreed!

>In the case where I set a max version in the labels, the intention is for the version to be no longer installable on OCP versions after that max version. However, it may still be able to run and upgrade on OCP versions after that max version - if it cannot maybe this is where deprecation comes into play (given that the resolver has access to that information via properties on the bundle, it could surface that somehow). To be honest we haven't explicitly made that distinction in the pipeline but it definitely feels like a reasonable ask given this enhancement. What do you think?

I believe that you are suggesting that we could prevent cluster upgrades based on whether or not the CSV has been deprecated. This is certainly possible, but I am concerned that:
- Operator Authors would be forced to mark a number of CSVs as deprecated before each OpenShift release to prevent cluster upgrades. Missing this step could allow Cluster Upgrades that disrupt operator services. Note that the workflow above does not include this downside.
- Upgrades would only be disabled if the index on the cluster is at the earliest version that includes all the deprecation changes.. 

If you want to pull in eyes from the pipeline team to discuss please do, I welcome any and all reviews :)

Reply:
I think we should plan to make use of declarative index for this.

- A team publishes an operator v1 during openshift 4.6. It is tested on 4.6 and stamped with maxOCPVersion: 4.6
- 4.7 is released. Some combination of QE / automated tests / static analysis determines to some degree of confidence that the operator should or does run on 4.7. 
- The index is updated to overwrite `maxOCPVersion` to be 4.7 for v1.

The goal of the declarative config is to allow us to overwrite metadata (like maxOCPVersion) for operators after release.

Certainly what @awgreene said will allow us to make use of maxOCPVersion without declarative config - but we know that we want to deliver declarative config for 4.8 already.

Reply:
@ecordell are you proposing that adopting declarative config is the only way operator authors will be able to specify maxVersion?

Reply:
Can we just include a bit about dropping properties into the bundle and making sure that this workflow works that way as well?

Reply:
I will bump the `Declaring Max OpenShift Version via the Declarative Index Config` section into the main proposal and outline the advantages that @ecordell discussed.

Reply:
> are you proposing that adopting declarative config is the only way operator authors will be able to specify maxVersion?

Nope - just that with the declarative config, there is no need to publish a _new_ bundle to support a new ocp versions. Existing bundles can be annotated with new support statements as we discover them.

Reply:
> Nope - just that with the declarative config, there is no need to publish a new bundle to support a new ocp versions. Existing bundles can be annotated with new support statements as we discover them.

ack, ok.

Reply:
Updated this portion of the document to outline the benefits of declaring the operators Max OpenShift Version using the Declarative Index Config.